test/externalTests/prb-math.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+
+# ------------------------------------------------------------------------------
+# This file is part of solidity.
+#
+# solidity is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2023 solidity contributors.
+# ------------------------------------------------------------------------------
+
+import sys
+import subprocess
+import json
+from pathlib import Path
+from shutil import copyfile
+
+from external_test import TestConfig, ExternalTest, CURRENT_EVM_VERSION
+from foundry import FoundryRunner
+
+class PRBMath(ExternalTest):
+    """PRBMath test setup"""
+
+    def __init__(self, config: TestConfig):
+        ExternalTest.__init__(self, config)
+
+    def setup_fn(self, test_dir: Path, env: dict):
+        """ PRBMath setup steps"""
+        copyfile(test_dir / ".env.example", test_dir / ".env")
+
+    def compile_fn(self, test_dir: Path, env: dict):
+        """ PRBMath compilation steps"""
+        # TODO
+
+    def test_fn(self, test_dir: Path, env: dict):
+        """ PRBMath tests steps"""
+        # TODO
+
+
+if __name__ == '__main__':
+    try:
+        # Note: We currently accept solc_binary_type and solc_binary_path as parameters
+        # to keep compatibility with the current external tests framework.
+        args = ExternalTest.parse_command_line(
+            "PRBMath external tests", sys.argv[1:])
+
+        config_json = json.loads(f"""
```

Comment:
I think that it would be good to redesign this part so that we have a single runner script that imports scripts for specific tests and allows running them by name. Then we could keep more repetitive stuff out of them.

The script could also let you run multiple tests (or all of them), making `externalTests.sh` obsolete.

Reply:
Right, yeah, my idea was indeed to replace `externalTests.sh` eventually. In fact, this script is not even used. The CI just iterates over each external test shellscript directly, see here: https://github.com/ethereum/solidity/blob/develop/.circleci/config.yml#L1245

Should I go for this redesign here in this PR and update the CI accordingly? i.e. making the CI invoke this new script (let's call it `externalTest.py` for now), passing the name of the external tests that should run. And as a temporary change, we can invoke the current shellscript for the tests that were not migrated yet.

Reply:
I'd do already it in this PR because it's not such a big change and it will actually save you some work (i.e. less boilerplate to repeat in the other test files).

The thing I'd do separately is actually the foundry support. I.e. first convert existing tests and only do Foundry as a PR on top of it.

Reply:
> I think that it would be good to redesign this part so that we have a single runner script that imports scripts for specific tests and allows running them by name. Then we could keep more repetitive stuff out of them.
> 
> The script could also let you run multiple tests (or all of them), making `externalTests.sh` obsolete.

On second thought, I think a single-runner script that imports other would not work very well for us here. Because many external projects rely on different framework versions (i.e. nodejs versions `14.20`, `16.18` and `18.11`, rust, etc), we currently use circleCI `cimgs` to install such dependencies. Of course we could also change that, but it would require to perform the installation in the script as well. And the `externalTests.sh` is not being used, is it?

I would rather keep the CI calling each external project script directly, like it is done now, and just migrate them to python using the common setup module we are creating.

Reply:
> Because many external projects rely on different framework versions (i.e. nodejs versions `14.20`, `16.18` and `18.11`, rust, etc), we currently use circleCI `cimgs` to install such dependencies. Of course we could also change that, but it would require to perform the installation in the script as well.

Don't worry about this, `externalTests.sh` can't deal with that either. It will currently just run all tests and if some of them do not work on your system, well, tough luck. Adding the runner actually helps because the idea is for it to allow running a subset of tests - so you can choose the ones you know will work for you. Currently the script is hard-coded to run them all.

> And the `externalTests.sh` is not being used, is it?

It might have been used in the distant past but right now it's not used in CI and there's no reason to. I think that having a separate jobs for each test is better.

And yeah, with current node versions it will likely not finish successfully if you run it locally. But these change and there were times where you could. Still, it's only mildly useful and I'd be fine with removing it if it were to stay in th current form. If we rewrite it to act as the runner though, it will become much more useful.

> I would rather keep the CI calling each external project script directly, like it is done now, and just migrate them to python using the common setup module we are creating.

Yes, that's what we should keep doing. Treat the ability to run multiple tests via the runner just as a convenience feature for local use.

Reply:
I made another PR to address your suggestion: https://github.com/ethereum/solidity/pull/14064 and I will remove the code duplication from this PR (i.e. `run_cmd` function) and also undo the changes on `externalTests.sh`.